### PR TITLE
Trainer Party Pools

### PR DESCRIPTION
**Pull request by hedara90**
## Description

 Allows trainers to generate their party from a curated pool of pokemon defined in the trainer party.
 This feature changes nothing about `trainers.party` or `trainers.h` if it's not used.
 To use this feature, simply define a `Party Size` that lower than the amount of defined pokemon for the trainer in `trainers.party` or `.partySize` and `.poolSize` in `traners.h` where `poolSize` is larger than `partySize` and matches the number of defined pokemon.
Optionally `Pool Rules` can be defined for the trainer to change which rules should be applied to the trainer.
 
Mons in the pool can have tags that change how the party is picked from the pool.
If a mon has the tag `Lead` the first mon in the party will be picked from all the pool members with this tag.
If a mon has the tag `Ace` the last member of the party will be picked from all the pool members with this tag.
Other tags can be marked as `POOL_TAG_REQUIRED` in `struct PoolRules poolRules` which will force that tag to be included as early as possible. (Not completely implemented right now)

Rules for the pools also include forcing unique species or NatDex numbers and unique items.
Items can also be excluded from the item uniqueness with a list.

Party generation is limited to a maximum size of 255 since index 255 is used for "No mon chosen yet".

## Images

trainerPool

## Things to note in the release changelog:

- Documentation on how to use this can be found under `docs/tutorials/how_to_trainer_party_pool.md`

## **Discord contact info**

hedara

### CODE REVIEW

File: tools/trainerproc/main.c [hunk 1]
```diff
@@ -16,8 +16,9 @@
 
 #define MAX_TRAINER_AI_FLAGS 32
 #define MAX_TRAINER_ITEMS 4
-#define PARTY_SIZE 6
+#define PARTY_SIZE 64
```
Comment (AlexOn1ine) [inline tools/trainerproc/main.c:19]: can we add some prefix to party_size? I understand it is in preproc but I would be surprised why this constant has 64 when we search for it.

Comment (hedara90) [reply tools/trainerproc/main.c:19]: The limit of 64 is right now arbitrary, the "true" limit is 255.
But a better name wouldn't hurt.

File: src/battle_main.c [hunk 2]
```diff
@@ -1936,8 +2212,21 @@ u8 CreateNPCTrainerPartyFromTrainer(struct Pokemon *party, const struct Trainer
             monsCount = trainer->partySize;
         }
 
+        bool32 usingPool = FALSE;
+        u8 *poolIndexArray;
```
Comment (fakuzatsu) [inline src/battle_main.c:2216]: I think this should be initialized to NULL. If nothing else it helps people who may still have `-werror=maybe-uninitialized` on, or who want to turn it on.

Comment (hedara90) [reply src/battle_main.c:2216]: Isn't it already on?
If I try to add a usage of it that I know is uninitialized it will warn me, but not with the current setup.

Comment (hedara90) [reply src/battle_main.c:2216]: 20241205_20h27m01s_grim

This is if I add a known uninitialized usage, the compiler seems smart enough to see that it can't be used uninitialized currently because its usage depends on a value being true that's set in the same function it's initialized in.

Comment (fakuzatsu) [reply src/battle_main.c:2216]: Weird. Perhaps it was my version of gcc, but I was getting `-Werror=maybe-uninitialized` with the current setup.

Comment (fakuzatsu) [reply src/battle_main.c:2216]: gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 btw

Comment (hedara90) [reply src/battle_main.c:2216]: gcc (GCC) 14.2.1 20240910
Does not give an error.

Comment (mrgriffin) [reply src/battle_main.c:2216]: > Weird. Perhaps it was my version of gcc, but I was getting `-Werror=maybe-uninitialized` with the current setup.

It probably is—GCC gets more advanced static analysis with every version :)

~~SBird will swing by in a minute to say that we should ensure we support older GCCs, like in #5611~~

Comment (fakuzatsu) [reply src/battle_main.c:2216]: There's no harm in initializing early, so may as well

Comment (hedara90) [reply src/battle_main.c]: Added the `NULL`.
(And some other fixes)

**Comment by hedara90** [comment]
Preliminarily ready for review.
Things left to do:

- Tests
- Implement pool verification

Replacement trainer for Tiana for testing.
```
=== TRAINER_TIANA ===
Name: TIANA
Class: Lass
Pic: Lass
Gender: Female
Music: Female
Double Battle: Yes
AI: Check Bad Move
Party Size: 3
Pool Rules: Weather Doubles

Zigzagoon
Level: 4
IVs: 0 HP / 0 Atk / 0 Def / 0 SpA / 0 SpD / 0 Spe

Shroomish
Level: 4
IVs: 0 HP / 0 Atk / 0 Def / 0 SpA / 0 SpD / 0 Spe

Psyduck
Level: 4
IVs: 0 HP / 0 Atk / 0 Def / 0 SpA / 0 SpD / 0 Spe

Shellder
Level: 4
IVs: 0 HP / 0 Atk / 0 Def / 0 SpA / 0 SpD / 0 Spe

Mew
Level: 4
IVs: 0 HP / 0 Atk / 0 Def / 0 SpA / 0 SpD / 0 Spe
Tags: Ace

Giratina
Level: 4
IVs: 0 HP / 0 Atk / 0 Def / 0 SpA / 0 SpD / 0 Spe
Tags: Ace

Vulpix
Ability: Drought
Level: 4
Tags: Lead / Weather Setter

Torkoal
Ability: Drought
Level: 4
Tags: Lead / Weather Setter

Bulbasaur
Ability: Chlorophyll
Level: 4
Tags: Lead / Weather Abuser

Cherrim
Level: 4
Tags: Lead / Weather Abuser
```

**Comment by hedara90** [comment]
(Almost) ready for review, just missing pool verification.
Very much usable, assuming all pools are valid, worst case they can be padded with random mons.
Documentation is in [docs/tutorials/how_to_trainer_party_pool.md](https://github.com/rh-hideout/pokeemerald-expansion/pull/5731/files#diff-ca3e482ad9aa31659ee01d11cf4a8a218bfecf9e10b9e1cf0e71a21e78ce6ad3)

**Comment by hedara90** [comment]
Ready for review.
Do not merge before #5337

**Comment by hedara90** [comment]
Split all the pool functions into its own file.

**Comment by hedara90** [comment]
Conflicts resolved, #5337 merged.
TPP is ready for review

File: src/trainer_pools.c [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+    }
+    return monIndex;
+}
+
+static u32 DefaultAcePickFunction(const struct Trainer *trainer, u8 *poolIndexArray, u32 partyIndex, u32 monsCount, u32 battleTypeFlags, struct PoolRules *rules)
+{
+    u32 arrayIndex = 0;
+    u32 monIndex = POOL_SLOT_DISABLED;
+    //  monIndex is set to 255 if nothing has been chosen yet, this gives an upper limit on pool size of 255
+    if (monIndex == POOL_SLOT_DISABLED
```
Comment (AlexOn1ine) [inline src/trainer_pools.c:65]: if we set `monIndex` to `POOL_SLOT_DISABLED`?

Generally this if condition looks very cursed. Maybe rewrite it with early returns or place the or conditions in the same line

Comment (hedara90) [reply src/trainer_pools.c:65]: This is just a remnant of when `DefaultAcePickFunction` was just an  `if` statement in `PickMonFromPool`.
Will fix locally.

File: src/trainer_pools.c [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+    u32 monIndex = POOL_SLOT_DISABLED;
+    //  Pick Lead
+    if (monIndex == POOL_SLOT_DISABLED)
+        monIndex = pickFunctions.LeadFunction(trainer, poolIndexArray, partyIndex, monsCount, battleTypeFlags, rules);
+    //  Pick Ace
+    if (monIndex == POOL_SLOT_DISABLED)
+        monIndex = pickFunctions.AceFunction(trainer, poolIndexArray, partyIndex, monsCount, battleTypeFlags, rules);
+    //  If no mon has been found yet continue looking
+    if (monIndex == POOL_SLOT_DISABLED)
+        monIndex = pickFunctions.OtherFunction(trainer, poolIndexArray, partyIndex, monsCount, battleTypeFlags, rules);
```
Comment (AlexOn1ine) [inline src/trainer_pools.c:189]: monIndex in this case is always disabled right?

Comment (hedara90) [reply src/trainer_pools.c:189]: No, if either a lead or an ace has been found, `monIndex` is not disabled.

File: src/trainer_pools.c [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+    }
+    return monIndex;
+}
+
+static u32 DefaultOtherPickFunction(const struct Trainer *trainer, u8 *poolIndexArray, u32 partyIndex, u32 monsCount, u32 battleTypeFlags, struct PoolRules *rules)
+{
+    u32 arrayIndex = 0;
+    u32 monIndex = POOL_SLOT_DISABLED;
+    //  monIndex is set to 255 if nothing has been chosen yet, this gives an upper limit on pool size of 255
+    if (monIndex == POOL_SLOT_DISABLED)
```
Comment (AlexOn1ine) [inline src/trainer_pools.c:117]: monIndex is set to `POOL_SLOT_DISABLED` and checked in the next condition

Comment (hedara90) [reply src/trainer_pools.c:117]: Same as for the ace function.
Fixed locally.

File: src/trainer_pools.c [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+        for (u32 currIndex = 0; currIndex < trainer->poolSize; currIndex++)
+        {
+            if (poolIndexArray[currIndex] != POOL_SLOT_DISABLED
+             && !(trainer->party[poolIndexArray[currIndex]].tags & (1u << POOL_TAG_LEAD))
+             && !(trainer->party[poolIndexArray[currIndex]].tags & (1u << POOL_TAG_ACE)))
+            {
+                if (firstUnpickedIndex == POOL_SLOT_DISABLED)
+                    firstUnpickedIndex = currIndex;
+                //  Start from index 2, since lead and ace has special handling
+                for (u32 currTag = 2; currTag < POOL_NUM_TAGS; currTag++)
```
Comment (AlexOn1ine) [inline src/trainer_pools.c:131]: I believe this tag loop and the tag loop above look identical. Maybe a function would make sense in this case to remove duplicate code.

**Comment by hedara90** [comment]
Addressed review comments.

File: include/config/battle.h [hunk 7]
```diff
@@ -299,4 +299,12 @@
 #define B_ENEMY_MON_SHADOW_STYLE        GEN_3 // In Gen4+, all enemy Pokemon will have a shadow drawn beneath them.
                                               // Currently Gen4+ shadows don't properly work with Trainerslides
 
+#define B_POOL_SETTING_CONSISTENT_RNG       FALSE // If set to true, the same trainer will always generate the same pool on the same save file
+#define B_POOL_SETTING_USE_FIXED_SEED       FALSE // If set to true, will use the fixed seed defined in B_POOL_SETTING_FIXED_SEED
+#define B_POOL_SETTING_FIXED_SEED           0x1D4127 // "Random" number, unless a mistake was made, it's へだら in Emerald charmap which should spell he-da-ra
+#define B_POOL_RULE_SPECIES_CLAUSE          TRUE  // Only pick a single pokemon of a unique NatDex number
+#define B_POOL_RULE_EXCLUDE_FORMS           TRUE  // Exclude different forms from the Species Clause
+#define B_POOL_RULE_ITEM_CLAUSE             TRUE  // Only allow each item to be picked once
```
Comment (AlexOn1ine) [inline include/config/battle.h:307]: shouldn't all rules be set to false?

File: src/battle_main.c [hunk 8]
```diff
@@ -1919,6 +1920,7 @@ void CustomTrainerPartyAssignMoves(struct Pokemon *mon, const struct TrainerMon
 
 u8 CreateNPCTrainerPartyFromTrainer(struct Pokemon *party, const struct Trainer *trainer, bool32 firstTrainer, u32 battleTypeFlags)
 {
+
```
Comment (AlexOn1ine) [inline src/battle_main.c:1923]: redundant line

File: src/trainer_pools.c [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+    u32 seed;
+    if (B_POOL_SETTING_USE_FIXED_SEED)
+        seed = B_POOL_SETTING_FIXED_SEED;
+    else
+        seed = gSaveBlock2Ptr->playerTrainerId[0] + (gSaveBlock2Ptr->playerTrainerId[1] << 8) + (gSaveBlock2Ptr->playerTrainerId[2] << 16) + (gSaveBlock2Ptr->playerTrainerId[3] << 24);
+    seed ^= (u32)trainer;
+    return seed;
+}
+
+static void RandomizePoolIndices(const struct Trainer *trainer, u8 *poolIndexArray)
```
Comment (AlexOn1ine) [inline src/trainer_pools.c:246]: why do you want to randomize the pool indices?

File: src/trainer_pools.c [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+
+            for (u32 i = 0; i < monsCount; i++)
+            {
+                monIndices[i] = PickMonFromPool(trainer, poolIndexArray, i, monsCount, battleTypeFlags, &rules, pickFunctions);
+                //  If the slot doesn't have a proper value, the pool creation failed, fall back to normal mon pick process
+                if (monIndices[i] == POOL_SLOT_DISABLED)
+                {
+                    usingPool = FALSE;
+                    break;
+                }
```
Comment (AlexOn1ine) [inline src/trainer_pools.c:383]: I'm not sure if I understand this. Should the whole process be stopped if there isn't a proper value?

**Review by AlexOn1ine — COMMENTED** [review]
This is actually a really great system for randomization

File: src/trainer_pools.c (hunk 9 above)
Comment (hedara90) [reply src/trainer_pools.c:246]: So that instead of calling random every time the pool needs to pick a mon, it's all done upfront and you can just iterate over the index array to find a "random" index.

File: src/trainer_pools.c (hunk 10 above)
Comment (hedara90) [reply src/trainer_pools.c:383]: Yes, if a proper value wasn't found, it means that the pool picking process failed somehow.
So stop everything and fall back to the usual "pick first N mons in the trainer party array".

File: src/trainer_pools.c (hunk 9 above)
Comment (AlexOn1ine) [reply src/trainer_pools.c:246]: oh that is smart

File: src/trainer_pools.c (hunk 10 above)
Comment (AlexOn1ine) [reply src/trainer_pools.c:383]: what if the first mon was valid but the next mon wasn't valid?

Comment (hedara90) [reply src/trainer_pools.c:383]: Then I'm judging the entire pool to be invalid.
I think it's better to have a consistent behavior for the failure state rather than potentially picking the some mon twice when falling back.

Comment (AlexOn1ine) [reply src/trainer_pools.c:383]: makes sense

File: include/config/battle.h (hunk 7 above)
Comment (hedara90) [reply include/config/battle.h:307]: It's just a default set of rules.
I guess that if it should match the behavior of PWT/Battle Maison/Tree `B_POOL_RULE_EXCLUDE_FORMS` should be set to `FALSE`.

File: src/battle_main.c (hunk 8 above)
Comment (hedara90) [reply src/battle_main.c:1923]: Addressed locally.
~~Time to run CI for a whitespace chance...~~

File: include/config/battle.h (hunk 7 above)
Comment (AlexOn1ine) [reply include/config/battle.h:307]: shouldn't by default everything be allowed?

Comment (hedara90) [reply include/config/battle.h:307]: I don't care overly much, so I can set it all to `FALSE`.

Comment (AlexOn1ine) [reply include/config/battle.h:307]: ye, I think it is better. It should be the users decision on what to do with it

**Comment by hedara90** [comment]
Comments addressed

File: include/data.h [hunk 11]
```diff
@@ -94,6 +95,10 @@ struct Trainer
              u8 startingStatus:6;    // this trainer starts a battle with a given status. see include/constants/battle.h for values
     /*0x1F*/ u8 mugshotColor;
     /*0x20*/ u8 partySize;
+    /*0x21*/ u8 poolSize;
+    /*0x22*/ u8 poolRuleIndex;
+    /*0x23*/ u8 poolPickIndex;
+    /*0x24*/ u8 poolPruneIndex;
```
Comment (AlexOn1ine) [inline include/data.h:101]: both of those could be just 4 bits right, or even 3 with 2 padding bits?

Comment (hedara90) [reply include/data.h:101]: Depends on how many custom functions users add, and iirc it wouldn't change the size of the struct to to the 32-bit alignment.

**Review by AlexOn1ine — APPROVED** [review]
